core/src/main/java/io/questdb/cairo/wal/WalPurgeJob.java

```diff
@@ -583,4 +413,216 @@ protected boolean runSerially() {
+                    debugBuffer.put(')');
+                }
+
+                if (i < n - 1) {
+                    debugBuffer.put(',');
+                }
+            }
+
+            debugBuffer.put("], nextToApply=[");
+            for (int i = 0, n = nextToApplyKeys.size(); i < n; ++i) {
+                final int walId = (int) nextToApplyKeys.getQuick(i);
+                final int segmentId = nextToApply.get(walId);
+                debugBuffer.put('(').put(walId).put(',').put(segmentId).put(')');
+                if (i < n - 1) {
+                    debugBuffer.put(',');
+                }
+            }
+            debugBuffer.put(']');
+        }
+
+        public boolean hasOnDiskSegments() {
+            return onDiskWalIDSegmentIDPairs.size() != 0;
+        }
+
+        public boolean hasPendingTasks() {
+            return pendingWalIDSet.size() > 0;
+        }
+
+        public void reset(TableToken tableToken) {
+            this.tableToken = tableToken;
+            nextToApply.clear();
+            nextToApplyKeys.clear();
+            onDiskWalIDSegmentIDPairs.clear();
+            lockedWalIDSet.clear();
+            pendingWalIDSet.clear();
+        }
+
+        public void run() {
+            sortTracked();
+            accumDebugState();
+
+            for (int i = 0, n = onDiskWalIDSegmentIDPairs.size(); i < n; i++) {
+                final long diskPair = onDiskWalIDSegmentIDPairs.get(i);
+                final int walId = Numbers.decodeHighInt(diskPair);
+                final int segmentKey = Numbers.decodeLowInt(diskPair);
+                final int segmentId = decodeSegmentId(segmentKey);
+                final boolean hasPendingTasks = decodePendingTasks(segmentKey);
+                final boolean segmentLocked = decodeLocked(segmentKey);
+
+                if (hasPendingTasks) {
+                    continue;
+                }
+
+                final int nextToApplySegmentId = nextToApply.get(walId);  // -1 if not found
+
+                if (nextToApplySegmentId != -1) {
+                    // WAL is in use, there are committed transactions in Sequencer
+                    if (nextToApplySegmentId > segmentId) {
+                        // Segment already applied.
+                        logDebugInfo();
+                        deleter.deleteSegmentDirectory(walId, segmentId);
+                    }
+                } else if (!segmentLocked) {
+                    // If the sequencer doesn't track any unapplied changes for this walId and the walId is unlocked,
+                    // then we can delete the whole WAL directory, unless any of its segments track pending tasks.
+                    if (!lockedWalIDSet.contains(walId) && !pendingWalIDSet.contains(walId)) {
+                        logDebugInfo();
+                        deleter.deleteWalDirectory(walId);
+                    } else {
+                        // WAL is locked but the segment in it is not in outstanding commits, delete it.
```

Comment:
This comment is incorrect, WAL can be unlocked at this point.

Reply:
These are all old comments that got dragged around in the refactor. I'll audit all of 'em and make sure they're up to date. Thanks for spotting it!

Reply:
Reworked the comments.